### ISSUE TITLE
Bug? ofThreadChannel can have size of 0 when tryReceive continues

### ISSUE THREAD

**Issue by elliotwoods** [openframeworks/openFrameworks#4946]
Hey!

![image](https://cloud.githubusercontent.com/assets/328294/13312918/a7e80a4c-dbdd-11e5-9d1c-1a2360235b57.png)
I'm getting crashes on this line because queue is empty

I have stable operation when I'm using tryReceive without the timeout, but it crashes on 5% of calls when I use this function instead.
To note, only 1 thread is receiving (so it's not like i'm unlocking elsewhere after i've emptied the queue). Also perhaps note that multiple threads are sending, but that doesn't seem to be unsupported operation.

Perhaps condition.wait_for can just return true for other reasons?
Anyway, the fix here is to check again if the queue is empty before entering the scope with swap

**Comment by arturoc** [comment]
this structure is usually mpsc, multiple producer single consumer and i've really only tested it with spsc. 

can you try changing notify_all for notify_one in send? it seems like the condition would notify everyone, if there's multiple consumers then the queue could get emptied and when the next one tries to swap it crashes because it's been emptied already.

**Comment by elliotwoods** [comment]
to clarify, i'm multiple producer, single consumer
which is why this is confusing (i can't imagine a situation where a lock would be released in another thread apart from when we add to the queue, i.e. queue should never be empty after an unlock from another thread)

**Comment by elliotwoods** [comment]
testing here... if all the producers are turned off then it still crashes
so i presume `wait_for` returns true if the timeout has been hit

**Comment by arturoc** [comment]
that's strange, accoding to the docs wait_for can only return != timeout if it actually didn't timeout or if there was an exception but in that case the exception is propagated if i understand it correctly: http://www.cplusplus.com/reference/condition_variable/condition_variable/wait_for/

**Comment by elliotwoods** [comment]
right, and `cv_status::timeout == 1`, so we need to do `if(cv_status::timeout != condition.wait_for(...`

**Comment by elliotwoods** [comment]
hmm.. wait, that's there already. ignore me! :)

**Comment by arturoc** [comment]
:)

**Comment by elliotwoods** [comment]
ok back to this
having the same problem again
according the the specification, a spurious wake would cause `wait_for` to return `no_timeout` even if no notification was called.

it's common in the specified use of `wait_for` to always ensure that your own condition is met before continuing because of spurious waking

http://www.cplusplus.com/reference/condition_variable/condition_variable/wait_for/

so the only safe implementation would be something like:

```c++
	bool tryReceive(T & sentValue, int64_t timeoutMs){
		std::unique_lock<std::mutex> lock(mutex);
		if(closed){
			return false;
		}
		if(queue.empty()){
			condition.wait_for(lock, std::chrono::milliseconds(timeoutMs));
			if(queue.empty()) {
				return false;
			}
		}

		if(!closed){
			std::swap(sentValue,queue.front());
			queue.pop();
			return true;
		}else{
			return false;
		}
	}
```

(rather than the original):
```c++
	bool tryReceive(T & sentValue, int64_t timeoutMs){
		std::unique_lock<std::mutex> lock(mutex);
		if(closed){
			return false;
		}
		if(queue.empty()){
			if(condition.wait_for(lock,std::chrono::milliseconds(timeoutMs))==std::cv_status::timeout){
				return false;
			}
		}

		if(!closed){
			std::swap(sentValue,queue.front());
			queue.pop();
			return true;
		}else{
			return false;
		}
	}
```

**Comment by elliotwoods** [comment]
note that spurious waking happens differently in different implementations
i'm finding it's happening on VS2015/x64 
perhaps it just never happens on your main test platforms?

**Comment by arturoc** [comment]
yeah i don't understand that very well, i'm using this class in several projects and never found any problem. 

I guess it depends on what it uses to wait and some signal other than the expected one wakes it. 

one thing we could do is calculate an offset in time from now and then wait_until instead of for. that way if the condition wakes before the timeout we can still wait using the same offset until at some point it returns true.

haven't tested but something like:

```cpp
bool tryReceive(T & sentValue, int64_t timeoutMs){
        std::unique_lock<std::mutex> lock(mutex);
        if(closed){
            return false;
        }

        auto until = std::chrono::steady_clock::now() + std::chrono::milliseconds(timeoutMs);
        auto ret = std::cv_status::no_timeout;
        while(queue.empty() && ret!=std::cv_status::timeout){
            ret = condition.wait_until(lock,until);
        }
        if(queue.empty()){
            return false;
        }

        if(!closed){
            std::swap(sentValue,queue.front());
            queue.pop();
            return true;
        }else{
            return false;
        }
    }
```

**Comment by elliotwoods** [comment]
ok, so guarantee that we wait at least that long to the user? makes sense
that would work for me over here!
i'll use another patched ThreadChannel class over here for now

**Comment by arturoc** [comment]
yes that's it, but what you sent would be ok too i think. if you can test it? my version has a while loop so if the end condition is not exactly as we think it might enter into an infinte loop.

can you send a PR for whatever works for you?
